gulpfile.ts

```diff
@@ -0,0 +1,159 @@
+import gulp from 'gulp';
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const clean = require('gulp-clean');
+import merge from 'merge-stream';
+import ts from 'gulp-typescript';
+import sourcemaps from 'gulp-sourcemaps';
+import terser from 'gulp-terser';
+import uglify from 'gulp-uglify';
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const rollup = require('gulp-rollup-2').rollup;
+import rollupTypescript from 'rollup-plugin-typescript2';
+
+import minifyPrivatesTransformer from 'ts-transformer-minify-privates';
+import tsMacroTransformer from 'ts-macros';
+import pathsTransformer from 'ts-transform-paths';
+import { CustomTransformerFactory, Program } from 'typescript';
+
+function createProject(overrideSettings?: ts.Settings) {
+  return ts.createProject('tsconfig.json', {
+    getCustomTransformers: (program?: Program) => {
+      if (program === undefined) throw new Error('Program is undefined');
+      return {
+        ...pathsTransformer(program),
+        before: [
+          tsMacroTransformer(program) as unknown as CustomTransformerFactory,
+          minifyPrivatesTransformer(program)
+        ]
+      };
+    },
+    ...overrideSettings
+  });
+}
+
+function terserStream() {
+  return terser({
+    compress: false,
+    output: {
+      beautify: true
+    },
+    mangle: {
+      keep_fnames: true,
+      keep_classnames: true,
+      properties: {
+        regex: /^_private_/
```

Comment:
Will this configuration take effect?

As far as I know, only attributes that are successfully matched by the regex will be replaced.

Maybe we shoule change all the `protected` and `private` properties name from 'xxx' to '_xxx'?

Reply:
Transforms names only for private members

The reason for doing this is that there is no information left about what is a private member after compilation, so if you do not mangle this part during deployment, you will not be able to apply this optimization later.

For example, if you use a bundler such as webpack for a project that uses js-sdsl to reduce the code size, the Bundler cannot change the property name short

Reply:
I see this in the output files:

```javascript
TreeContainer.prototype.insertNodeSelfBalance = ...
```

It is a protected member but also is a private member.

Did you mean we cannot transformer `protected` name to be short?

Reply:
Yes, protected members cannot be mangled. Because you can also access protected members from outside the library.

For example, you can inherit the HashContainer from outside the library to create a new container. That's why protected members can't do mangling.